network/wsNetwork.go

```diff
@@ -1286,7 +1287,11 @@ func (wn *WebsocketNetwork) messageHandlerThread(peersConnectivityCheckCh <-chan
 			switch outmsg.Action {
 			case Disconnect:
 				wn.wg.Add(1)
-				go wn.disconnectThread(msg.Sender, disconnectBadData)
+				reason := disconnectBadData
+				if outmsg.reason != disconnectReasonNone {
+					reason = outmsg.reason
+				}
+				go wn.disconnectThread(msg.Sender, reason)
```

Comment:
clever! I was thinking about the work required to include the badData to all disconnect messages, but this is way cleaner.